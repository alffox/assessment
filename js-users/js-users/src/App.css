span.switch-handle.btn.btn-lightbtn-sm {
    background-color: #fff;
}