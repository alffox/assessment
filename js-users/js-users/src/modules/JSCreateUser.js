import React from "react";

import { Row, Col, Form, Button } from 'react-bootstrap';

class JSCreateUser extends React.Component {

    handleNewFirstName(e) {
        this.setState({
            new_first_name: e.target.value
        });
    }

    handleNewLastName(e) {
        this.setState({
            new_last_name: e.target.value
        });
    }


    handleNewUserFormSubmission = () => {
        fetch('http://js-assessment-backend.herokuapp.com/users.json', {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                first_name: this.state.new_first_name,
                last_name: this.state.new_last_name,
                status: "active"
            })
        })
            .then(function (response) {
                response.text().then((s) => console.log((s)));
            }).catch(function (error) {
                console.log(error.body);
            });
    };

    render() {
        return (
            <div>
                <br />
                <Form>

                    <Form.Group controlId="first-name">
                        <Form.Control type="input" placeholder="Enter User's First Name" defaultValue={this.props.new_first_name} onChange={e => this.handleNewFirstName(e)} />
                    </Form.Group>

                    <Form.Group controlId="last-name">
                        <Form.Control type="text" placeholder="Enter User's Last Name" defaultValue={this.props.new_last_name} onChange={e => this.handleNewLastName(e)} />
                    </Form.Group>
                    <Button variant="dark" type="button" onClick={(e) => this.handleNewUserFormSubmission(e)}>
                        Submit
                    </Button>
                </Form>
                <br />
            </div>
        );
    }
}

export default JSCreateUser;