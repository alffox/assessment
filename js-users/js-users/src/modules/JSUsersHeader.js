import React from "react";

class JSUsersHeader extends React.Component {
    render() {
        return (
            <header>
                <h1 className="text-center">JS-Users App</h1>
            </header>
        );
    }
}

export default JSUsersHeader;