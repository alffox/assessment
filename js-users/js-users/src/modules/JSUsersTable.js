import React from "react";

import { Table, Pagination } from 'react-bootstrap';
import BootstrapSwitchButton from 'bootstrap-switch-button-react' // Docs @ https://gitbrent.github.io/bootstrap-switch-button-react/
import Moment from 'react-moment';


class JSUsersTable extends React.Component {

    render() {
        // Solution inspired by https://stackoverflow.com/questions/40232847/how-to-implement-pagination-in-reactjs
        const indexOfLastUser = this.props.currentPage * this.props.usersPerPage;
        const indexOfFirstUser = indexOfLastUser - this.props.usersPerPage;
        const currentUsers = this.props.users.slice(indexOfFirstUser, indexOfLastUser);

        const pageNumbers = [];
        for (let i = 1; i <= Math.ceil(this.props.users.length / this.props.usersPerPage); i++) {
            pageNumbers.push(i);
        }

        const renderPageNumbers = pageNumbers.map(number => {
            return (
                <Pagination.Item
                    key={number}
                    id={number}
                    active={number === this.props.currentPage}
                    onClick={this.props.handlePaginatorClick.bind(this)}
                >
                    {number}
                </Pagination.Item>
            );
        });
        return (
            <div>
                <Table striped bordered hover responsive variant="dark">
                    <thead>
                        <tr>
                            <th>Status</th>
                            <th>First Name</th>
                            <th>Last Name</th>
                            <th>Created At</th>
                        </tr>
                    </thead>
                    <tbody>
                        {currentUsers.map((user, index) => {
                            return (
                                <tr key={index}>
                                    {user.status === "active" ? (
                                        <td>
                                            <BootstrapSwitchButton
                                                size="sm"
                                                width={75}
                                                checked={true}
                                                onlabel='Active'
                                                offlabel='Locked'
                                                onChange={this.props.toggleStatus.bind(this, user)}
                                                onstyle="light"
                                                offstyle="dark"
                                            />
                                        </td>
                                    ) : (
                                            <td>
                                                <BootstrapSwitchButton
                                                    size="sm"
                                                    width={75}
                                                    checked={false}
                                                    onlabel='Active'
                                                    offlabel='Locked'
                                                    onChange={this.props.toggleStatus.bind(this, user)}
                                                    onstyle="light"
                                                    offstyle="dark"
                                                />
                                            </td>
                                        )}
                                    {user.status === "active" ? (
                                        <td>{user.first_name}</td>
                                    ) : (
                                            <td><del>{user.first_name}</del></td>
                                        )}
                                    {user.status === "active" ? (
                                        <td>{user.last_name}</td>
                                    ) : (
                                            <td><del>{user.last_name}</del></td>
                                        )}
                                    {user.status === "active" ? (
                                        <td><Moment format="DD MMMM GGGG, hh:mm:ss">{user.created_at}</Moment></td>
                                    ) : (
                                            <td><del><Moment format="DD MMMM GGGG, hh:mm:ss">{user.created_at}</Moment></del></td>
                                        )}
                                </tr>
                            );
                        })}
                    </tbody>
                </Table>
                <Pagination size="small" className="d-flex flex-wrap">{renderPageNumbers}</Pagination>
            </div>
        );
    }
}

export default JSUsersTable;