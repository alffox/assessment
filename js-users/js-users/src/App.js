import React from 'react';
import './App.css';

import { Container, Row, Nav, Navbar, Image } from 'react-bootstrap';

import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";

import JSUsersError from "./modules/JSUsersError.js";
import JSUsersHeader from "./modules/JSUsersHeader.js";
import JSUsersTable from "./modules/JSUsersTable.js";
import JSCreateUser from "./modules/JSCreateUser.js";

class App extends React.Component {
  constructor() {
    super();
    this.state = {
      error: null,
      isLoading: false,
      users: [],
      currentPage: 1,
      usersPerPage: 10
    };
    this.handlePaginatorClick = this.handlePaginatorClick.bind(this);
    this.toggleStatus = this.toggleStatus.bind(this);
  }

  handlePaginatorClick(event) {
    this.setState({
      currentPage: Number(event.target.id)
    });
  }

  toggleStatus(user) {
    if (user.status === "active") {
      user.status = "locked";
      this.setState({ user: !this.state.user })
    } else if (user.status === "locked") {
      user.status = "active";
      this.setState({ user: !this.state.user })
    }
  }

  componentDidMount() {
    this.setState({ isLoading: true }, () => {
      fetch("http://js-assessment-backend.herokuapp.com/users.json")
        .then(res => res.json())
        .then(
          (data) => {
            this.setState({
              isLoading: false,
              users: data
            });
          },
          // Note: it's important to handle errors here
          // instead of a catch() block so that we don't swallow
          // exceptions from actual bugs in components.
          (error) => {
            this.setState({
              isLoading: true,
              error
            });
          });
    }
    )
  }

  render() {
    return (
      <Container>
        <JSUsersHeader />
        {this.state.error ? (
          <JSUsersError
            error={this.state.error} />
        ) : (
            <Router>
              <Navbar className="navbar navbar-dark bg-dark expand" expand="lg">
                <Navbar.Brand>JS-Users</Navbar.Brand>
                <Navbar.Toggle aria-controls="basic-navbar-nav" />
                <Navbar.Collapse id="basic-navbar-nav">
                  <Nav className="mr-auto">
                    <Nav.Link as={Link} to="" >Home</Nav.Link>
                    <Nav.Link as={Link} to="/new" >Create New User</Nav.Link>
                  </Nav>
                </Navbar.Collapse>
              </Navbar>

              <Switch>
                <Route path="/new" render={(props) =>
                  <JSCreateUser {...props}
                  />}
                />
              </Switch>
              <Switch>
                <Route path="" render={(props) =>
                  <JSUsersTable {...props}
                    isLoading={this.state.isLoading}
                    users={this.state.users}
                    currentPage={this.state.currentPage}
                    usersPerPage={this.state.usersPerPage}
                    toggleStatus={this.toggleStatus}
                    handlePaginatorClick={this.handlePaginatorClick} />}
                />
              </Switch>
            </Router>
          )}
        <small>Icons made by <a href="https://www.flaticon.com/authors/smashicons" title="Smashicons">Smashicons</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a></small>
      </Container>
    );
  }
}
export default App;
