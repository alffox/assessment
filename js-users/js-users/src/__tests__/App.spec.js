import React from 'react';
import renderer from 'react-test-renderer';

import App from '../App.js';
import JSUsersHeader from "../modules/JSUsersHeader.js";
import JSUsersTable from "../modules/JSUsersTable.js";
import JSCreateUser from "../modules/JSCreateUser.js";


describe('App', () => {
    test('snapshot renders', () => {
        const component = renderer.create(<App />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('JSUsersHeader', () => {
    test('snapshot renders', () => {
        const component = renderer.create(<JSUsersHeader />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('JSUsersTable', () => {
    test('snapshot renders', () => {
        const component = renderer.create(<JSUsersTable />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('JSCreateUser', () => {
    test('snapshot renders', () => {
        const component = renderer.create(<JSCreateUser />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});