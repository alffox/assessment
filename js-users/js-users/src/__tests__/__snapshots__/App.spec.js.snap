// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div
  className="container"
>
  <header
    className="d-flex flex-wrap"
  >
    <img
      alt="Pencil Cup"
      className="logo"
      src="logo.svg"
    />
    <h1
      className="text-center mt-3"
    >
      JS-Users App
    </h1>
  </header>
  <nav
    className="navbar navbar-dark bg-dark expand navbar navbar-expand-lg navbar-light"
  >
    <span
      className="navbar-brand"
    >
      JS-Users
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="mr-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key={null}
          disabled={false}
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
        <a
          className="nav-link"
          data-rb-event-key={null}
          disabled={false}
          href="/new"
          onClick={[Function]}
        >
          Create New User
        </a>
      </div>
    </div>
  </nav>
  <div>
    <div
      className="text-center"
    >
      <div
        className="spinner-grow"
      />
      Loading data ...
    </div>
    <ul
      className="d-flex flex-wrap pagination pagination-small"
    />
  </div>
  <small>
    Icons made by 
    <a
      href="https://www.flaticon.com/authors/smashicons"
      title="Smashicons"
    >
      Smashicons
    </a>
     from 
    <a
      href="https://www.flaticon.com/"
      title="Flaticon"
    >
      www.flaticon.com
    </a>
  </small>
</div>
`;

exports[`JSCreateUser snapshot renders 1`] = `
<div>
  <br />
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            defaultValue=""
            id="first-name"
            onChange={[Function]}
            placeholder="First Name"
            type="input"
          />
        </div>
      </div>
      <div />
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            defaultValue=""
            id="last-name"
            onChange={[Function]}
            placeholder="Last Name"
            sm="3"
            type="text"
          />
        </div>
      </div>
      <div />
    </div>
    <button
      className="btn btn-dark"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
    <hr />
    <div />
  </form>
</div>
`;

exports[`JSUsersHeader snapshot renders 1`] = `
<header
  className="d-flex flex-wrap"
>
  <img
    alt="Pencil Cup"
    className="logo"
    src="logo.svg"
  />
  <h1
    className="text-center mt-3"
  >
    JS-Users App
  </h1>
</header>
`;
